File: audiobook_transfer/parser.py
"""Filename parser for audiobook naming patterns.
Combines regex patterns and heuristic matching from both projects.
"""
import re
from typing import Optional
from .models import ParsedInfo
from .utils import normalize_author, logger

# Regex patterns for structured audiobook naming conventions.
PATTERNS = [
    # Author - Series, Book N - Title
    (r'^(.+?)\s*[-–—]\s*(.+?),\s*Book\s*([\d.]+)\s*[-–—]\s*(.+)$',
     {'author': 1, 'series': 2, 'series_position': 3, 'title': 4, 'confidence': 90}),

    # Author - Series, Book N
    (r'^(.+?)\s*[-–—]\s*(.+?),\s*Book\s*([\d.]+)$',
     {'author': 1, 'series': 2, 'series_position': 3, 'confidence': 80}),

    # Author - Title [ASIN]
    (r'^(.+?)\s*[-–—]\s*(.+?)\s*\[([A-Z0-9]{10})\]$',
     {'author': 1, 'title': 2, 'asin': 3, 'confidence': 85}),

    # Word NN - Title (series with position, e.g. "Pern 01 - Dragonflight")
    (r'^([A-Za-z]+)\s+(\d+(?:\.\d+)?)\s+[-–—]\s+(.+)$',
     {'series': 1, 'series_position': 2, 'title': 3, 'confidence': 80}),

    # Author - Title (standard, lower confidence; requires space around dash)
    (r'^(.+?)\s+[-–—]\s+(.+)$',
     {'author': 1, 'title': 2, 'confidence': 70}),

    # [NN] Title (numbered series entry)
    (r'^\[(\d+)\]\s*(.+)$',
     {'series_position': 1, 'title': 2, 'confidence': 60}),

    # NN Title (no brackets, numbered series entry)
    (r'^(\d+)\s+(.+)$',
     {'series_position': 1, 'title': 2, 'confidence': 60}),

    # Title [ASIN]
    (r'^(.+?)\s*\[([A-Z0-9]{10})\]$',
     {'title': 1, 'asin': 2, 'confidence': 65}),

    # Just title (catch-all)
    (r'^(.+)$',
     {'title': 1, 'confidence': 30}),
]

# Words that indicate the string is NOT an author name.
NON_AUTHOR_WORDS = [
    'gothic horror', 'chapterized', 'series', 'unabridged', 'audiobook',
    'radio drama', 'book', 'entangled with fae', 'beauty and the beast',
    'retelling', 'abridged', 'dramatized', 'omnibus', 'collection',
    'edition', 'version', 'narrated', 'vol.', 'volume', 'trilogy',
    'saga', 'chronicles', 'cycle', 'anthology',
]

# Title indicator words for reverse pattern detection.
TITLE_WORDS = {'the', 'a', 'an', 'of', 'and', 'in', 'on', 'at', 'to',
               'for', 'with', 'from', 'by', 'or', 'it', 'is', 'no', 'not'}

AUDIO_EXTS = ['.m4b', '.m4a', '.mp3', '.aax', '.ogg', '.wma', '.flac', '.wav', '.aac', '.m4p', '.zip']
IMAGE_EXTS = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp']


def strip_all_parens(s: str) -> str:
    """Remove all parenthetical groups from a string."""
    while True:
        start = s.find('(')
        if start < 0:
            break
        end = s.find(')', start)
        if end < 0:
            break
        s = s[:start] + s[end + 1:]
    return ' '.join(s.split())


def is_authorish(s: str) -> bool:
    """Heuristic: does this string look like a person's name?"""
    s = s.strip()
    if not s:
        return False
    words = s.split()
    if len(words) < 1 or len(words) > 5:
        return False
    for w in words:
        if len(w) > 15:
            return False
    lower = s.lower()
    for na in NON_AUTHOR_WORDS:
        if na in lower:
            return False
    return True


def is_author_name(s: str) -> bool:
    """Stricter check: is this string DEFINITELY an author name?
    Used for reverse pattern detection (Title - Author).
    """
    s = s.strip()
    if not s:
        return False
    words = s.split()
    if len(words) < 1 or len(words) > 4:
        return False
    lower = s.lower()
    if lower.startswith('the ') or lower.startswith('a ') or lower.startswith('an '):
        return False
    # Reject strings containing digits (not author names)
    for w in words:
        for c in w:
            if c.isdigit():
                return False
    title_word_count = sum(1 for w in words if w.lower() in TITLE_WORDS)
    if len(words) > 2 and title_word_count / len(words) > 0.3:
        return False
    for w in words:
        if len(w) > 15:
            return False
    return True


def is_title_like(s: str) -> bool:
    """Check if a string looks like a book title (not an author)."""
    s = s.strip()
    if not s:
        return False
    lower = s.lower()
    # Titles often start with articles
    if lower.startswith('the ') or lower.startswith('a ') or lower.startswith('an '):
        return True
    # Common title keywords
    title_keywords = [
        'chronicles', 'saga', 'trilogy', 'cycle', 'series',
        'of ', 'and ', 'in ', 'on ', 'at ', 'the ',
    ]
    keyword_count = sum(1 for kw in title_keywords if kw in lower)
    words = s.split()
    # Multi-word strings with many small words are likely titles
    if len(words) >= 4:
        return True
    # 3-word strings with at least one preposition/article likely title
    if len(words) == 3 and keyword_count >= 1:
        return True
    return keyword_count >= 2


def heuristic_parse(clean: str, info: ParsedInfo):
    """Heuristic parsing for non-standard patterns that regex misses.

    Handles:
    - "Series Name (Author)" — author in parenthetical
    - "Title - Author" — reverse pattern detection
    """
    # Pattern: "Series Name (Author)" — author in last parenthetical
    last_open = clean.rfind('(')
    if last_open >= 0:
        last_close = clean.rfind(')')
        if last_close > last_open:
            before = clean[:last_open].strip()
            paren_content = clean[last_open + 1:last_close].strip()

            before = strip_all_parens(before)

            if ' - ' not in before and is_authorish(paren_content):
                info.author = paren_content
                info.series = before
                info.confidence = max(info.confidence, 75)
                return
            else:
                clean = before
        else:
            clean = strip_all_parens(clean)
    else:
        clean = strip_all_parens(clean)

    # Pattern: "Author - Title" or "Title - Author"
    if ' - ' in clean:
        idx = clean.index(' - ')
        left = clean[:idx].strip()
        right = clean[idx + 3:].strip()

        # Check for reverse pattern (Title - Author)
        if is_author_name(right) and not is_author_name(left):
            info.author = right
            info.title = left
            info.confidence = max(info.confidence, 70)
            return

        # Standard: Author - Title
        info.author = left
        info.title = strip_all_parens(right)
        info.confidence = max(info.confidence, 65)
        return

    # Fallback: entire string as title
    if not info.title:
        info.title = clean
        if info.confidence < 30:
            info.confidence = 30


def regex_parse(clean: str, info: ParsedInfo):
    """Try structured regex patterns against the clean name."""
    for pattern, mapping in PATTERNS:
        match = re.match(pattern, clean, re.DOTALL)
        if not match:
            continue

        groups = match.groups()
        regex_conf = mapping.get('confidence', 0)

        # Only override heuristic results if regex confidence is higher
        if regex_conf > info.confidence:
            info.title = None
            info.author = None
            info.series = None
            info.series_position = None

        for field_name, group_idx in mapping.items():
            if field_name == 'confidence':
                info.confidence = max(info.confidence, regex_conf)
            elif group_idx - 1 < len(groups):
                value = groups[group_idx - 1].strip()
                if field_name == 'series_position':
                    try:
                        info.series_position = float(value)
                    except ValueError:
                        pass
                elif field_name == 'year':
                    try:
                        info.year = int(value)
                    except ValueError:
                        pass
                elif field_name == 'author':
                    if not info.author:
                        info.author = value
                elif field_name == 'title':
                    if not info.title:
                        info.title = value
                elif field_name == 'series':
                    if not info.series:
                        info.series = value
                elif field_name == 'asin':
                    if not info.asin:
                        info.asin = value
        break


def parse_parent_context(info: ParsedInfo, parent_name: Optional[str]):
    """Extract series name and author from parent directory name."""
    if not parent_name:
        return

    parent_clean = strip_all_parens(parent_name)

    # Try "Author - ..." pattern from parent for author
    if not info.author:
        m = re.match(r'^([^-–—]+?)\s*[-–—]\s+', parent_clean)
        if m:
            info.author = normalize_author(m.group(1).strip())
            info.confidence = max(info.confidence, 40)
            logger.debug(f"  Author from parent: '{info.author}'")

    # Try "Series (Author)" pattern from parent
    if not info.series or not info.author:
        last_open = parent_name.rfind('(')
        if last_open >= 0:
            last_close = parent_name.rfind(')')
            if last_close > last_open:
                before = parent_name[:last_open].strip()
                paren = parent_name[last_open + 1:last_close].strip()
                if is_authorish(paren) and ' - ' not in before:
                    if not info.author:
                        info.author = normalize_author(paren)
                        info.confidence = max(info.confidence, 50)
                    if not info.series and before:
                        info.series = before
                        info.confidence = max(info.confidence, 50)

    # Look for series keywords in parent name
    keyword_match = re.search(
        r'\s+(Series|Saga|Trilogy|Cycle|Chronicles|books\s+\d+)\b',
        parent_name, re.IGNORECASE
    )

    if keyword_match and not info.series:
        before = parent_name[:keyword_match.start()].strip()
        words = before.split()

        if words:
            # Don't extract author from parent if before starts with The/A/An
            lower_before = before.lower()
            if not info.author and len(words) >= 3 and len(words) <= 5 \
                    and not lower_before.startswith('the ') \
                    and not lower_before.startswith('a ') \
                    and not lower_before.startswith('an '):
                potential_author = ' '.join(words[:2])
                if is_author_name(potential_author):
                    info.author = potential_author
                    info.confidence = max(info.confidence, 35)
                    info.series = ' '.join(words[2:])
                    info.confidence = max(info.confidence, 40)
                    return

            # Fallback: use entire before as series
            info.series = before
            info.confidence = max(info.confidence, 40)
            logger.debug(f"  Series from parent: '{info.series}'")


def extract_asin(clean: str, info: ParsedInfo):
    """Find ASIN anywhere in the string."""
    if info.asin:
        return
    asin_match = re.search(r'\[([A-Z0-9]{10})\]', clean)
    if asin_match:
        info.asin = asin_match.group(1)
        info.confidence = max(info.confidence, 60)


def parse_name(name: str, parent_name: Optional[str] = None) -> ParsedInfo:
    """Parse an audiobook filename/directory name into metadata components.

    Uses a two-pass approach:
    1. Heuristic parsing (handles Series (Author), reverse patterns)
    2. Regex parsing (handles structured naming conventions)

    Args:
        name: The filename or directory name to parse
        parent_name: The parent directory name for additional context
    """
    clean = name.strip()

    # Strip audio/image extensions
    lower = clean.lower()
    for ext in AUDIO_EXTS + IMAGE_EXTS:
        if lower.endswith(ext):
            clean = clean[:-len(ext)]
            break

    info = ParsedInfo(raw_name=name)

    # Pass 1: Heuristic parsing
    heuristic_parse(clean, info)

    # Pass 2: Regex parsing (may override heuristic if higher confidence)
    regex_parse(clean, info)

    # Post-process: inherit author from parent if parent looks like author name
    if not info.author and parent_name:
        parent = parent_name
        author_to_use = parent_name
        if ',' in parent:
            parent = parent.split(',')[0].strip()
            author_to_use = parent
        if is_authorish(parent) and not is_title_like(parent):
            words = author_to_use.split()
            if len(words) <= 4:
                info.author = author_to_use
                info.confidence = max(info.confidence, 45)

    # Post-process: extract series position from Volume/Vol NN
    if info.series_position is None and info.title:
        vol_match = re.search(r'(?i)\s*(?:Volume|Vol\.?)\s*(\d+(?:\.\d+)?)', info.title)
        if vol_match:
            info.series_position = float(vol_match.group(1))
            info.title = (info.title[:vol_match.start()] + info.title[vol_match.end():]).strip()
            info.confidence = max(info.confidence, 55)
        else:
            bracket_match = re.search(r'(?i)\s*\[(?:Volume|Vol\.?)\s*(\d+(?:\.\d+))\]', info.title)
            if bracket_match:
                info.series_position = float(bracket_match.group(1))
                info.title = re.sub(r'(?i)\s*\[(?:Volume|Vol\.?)\s*\d+(?:\.\d+)\]', '', info.title).strip()
                info.confidence = max(info.confidence, 55)

    # Clean up title after volume removal
    if info.title:
        info.title = info.title.rstrip(' ,;.')
        info.title = re.sub(r'\[\s*\]', '', info.title).strip()
        info.title = re.sub(r'^\[[A-Z0-9]+\]\s*', '', info.title).strip()
        info.title = re.sub(r'(?i)\s*\((?:Audiobook|Unabridged|Unabr)\)', '', info.title).strip()
        info.title = re.sub(r'(?i)\s*\{[^}]*\}', '', info.title).strip()

    # Post-process: extract title from "Series NN - Title" pattern
    if not info.series and info.title:
        series_match = re.match(r'^(\w+)\s+(\d+(?:\.\d+)?)\s+[-–—]\s+(.+)$', info.title)
        if series_match:
            info.series = series_match.group(1)
            info.series_position = float(series_match.group(2))
            info.title = series_match.group(3)
            info.confidence = max(info.confidence, 65)

    # Post-processing
    if info.author:
        info.author = normalize_author(info.author)

    # Parent context
    parse_parent_context(info, parent_name)

    # ASIN extraction
    extract_asin(clean, info)

    logger.debug(f"Parsed '{name}' -> author={info.author}, title={info.title}, "
                 f"series={info.series}, pos={info.series_position}, conf={info.confidence}")
    return info

File: pkg/parser/parser.go
// Package parser provides filename parsing for audiobook naming patterns.
// Combines regex patterns and heuristic matching for maximum coverage.
package parser

import (
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/jevonx/audioTransfer/pkg/models"
	"github.com/jevonx/audioTransfer/pkg/utils"
)

// pattern represents a regex pattern with field mappings.
type pattern struct {
	re      *regexp.Regexp
	mapping map[string]int // field_name -> group_index (1-based)
}

// Regex patterns covering common audiobook naming conventions.
var regexPatterns = []pattern{
	// Author - Series, Book N - Title
	{regexp.MustCompile(`^(.+?)\s*[-–—]\s*(.+?),\s*Book\s*([\d.]+)\s*[-–—]\s*(.+)$`),
		map[string]int{"author": 1, "series": 2, "series_position": 3, "title": 4, "_conf": 90}},

	// Author - Series, Book N
	{regexp.MustCompile(`^(.+?)\s*[-–—]\s*(.+?),\s*Book\s*([\d.]+)$`),
		map[string]int{"author": 1, "series": 2, "series_position": 3, "_conf": 80}},

	// Author - Title [ASIN]
	{regexp.MustCompile(`^(.+?)\s*[-–—]\s*(.+?)\s*\[([A-Z0-9]{10})\]$`),
		map[string]int{"author": 1, "title": 2, "asin": 3, "_conf": 85}},

	// [NN] Title (numbered series entry)
	{regexp.MustCompile(`^\[(\d+)\]\s*(.+)$`),
		map[string]int{"series_position": 1, "title": 2, "_conf": 60}},

	// NN Title (no brackets)
	{regexp.MustCompile(`^(\d+)\s+(.+)$`),
		map[string]int{"series_position": 1, "title": 2, "_conf": 60}},

	// Title [ASIN]
	{regexp.MustCompile(`^(.+?)\s*\[([A-Z0-9]{10})\]$`),
		map[string]int{"title": 1, "asin": 2, "_conf": 65}},

	// Word NN - Title (series with position, e.g. "Pern 01 - Dragonflight")
	{regexp.MustCompile(`^([A-Za-z]+)\s+(\d+(?:\.\d+)?)\s+[-–—]\s+(.+)$`),
		map[string]int{"series": 1, "series_position": 2, "title": 3, "_conf": 80}},

	// Author - Title (standard, lower confidence; requires space around dash to avoid kebab-case)
	{regexp.MustCompile(`^(.+?)\s+[-–—]\s+(.+)$`),
		map[string]int{"author": 1, "title": 2, "_conf": 70}},
}

// Audio extensions to strip before parsing.
var audioExts = []string{".m4b", ".m4a", ".mp3", ".aax", ".ogg", ".wma", ".flac", ".wav", ".aac", ".m4p", ".zip"}

// Words that indicate the string is NOT an author name.
var nonAuthorWords = []string{
	"gothic horror", "chapterized", "series", "unabridged", "audiobook",
	"radio drama", "book", "entangled with fae", "beauty and the beast",
	"retelling", "abridged", "dramatized", "omnibus", "collection",
	"edition", "version", "narrated", "vol.", "volume", "trilogy",
	"saga", "chronicles", "cycle", "anthology",
}

// Title indicator words for reverse pattern detection.
var titleWords = map[string]bool{
	"the": true, "a": true, "an": true, "of": true, "and": true,
	"in": true, "on": true, "at": true, "to": true, "for": true,
	"with": true, "from": true, "by": true, "or": true, "it": true,
	"is": true, "no": true, "not": true,
}

// ParseName parses an audiobook filename/directory name into metadata.
// parentName provides context from the parent directory.
func ParseName(name string, parentName string) *models.ParsedInfo {
	clean := strings.TrimSpace(name)

	// Strip audio/image extensions
	lower := strings.ToLower(clean)
	for _, ext := range audioExts {
		if strings.HasSuffix(lower, ext) {
			clean = clean[:len(clean)-len(ext)]
			break
		}
	}
	// Also strip image extensions
	for ext := range utils.CoverExts() {
		if strings.HasSuffix(strings.ToLower(clean), ext) {
			clean = clean[:len(clean)-len(ext)]
			break
		}
	}

	info := &models.ParsedInfo{
		RawName: name,
		Extra:   make(map[string]string),
	}

	// First pass: try heuristic parsing (handles Series (Author), reverse patterns)
	heuristicParse(clean, info)

	// Second pass: if regex gives us more info, use it
	regexParse(clean, info)

	// Post-process: inherit author from parent if parent looks like author name
	if info.Author == "" && parentName != "" && !strings.Contains(parentName, "/") {
		parentClean := parentName
		authorToUse := parentName
		// Handle comma-separated multi-author: "Caroline Peckham, Susanne Valenti"
		// Use first author for inheritance check and as the author value
		if strings.Contains(parentClean, ",") {
			parentClean = strings.TrimSpace(strings.Split(parentClean, ",")[0])
			authorToUse = parentClean
		}
		if isAuthorish(parentClean) && !isTitleLike(parentClean) {
			words := strings.Fields(authorToUse)
			if len(words) <= 4 {
				info.Author = authorToUse
				if info.Confidence < 45 {
					info.Confidence = 45
				}
			}
		}
	}

	// Post-process: extract series position from Volume/Vol NN anywhere in title, remove from title
	if info.SeriesPosition == 0 && info.Title != "" {
		volRe := regexp.MustCompile(`(?i)\s*(?:Volume|Vol\.?)\s*(\d+(?:\.\d+)?)`)
		if loc := volRe.FindStringSubmatchIndex(info.Title); loc != nil {
			if v, err := strconv.ParseFloat(info.Title[loc[2]:loc[3]], 64); err == nil {
				info.SeriesPosition = v
				// Remove volume text from title
				info.Title = strings.TrimSpace(info.Title[:loc[0]] + info.Title[loc[1]:])
				if info.Confidence < 55 {
					info.Confidence = 55
				}
			}
		}
		// Also handle [Volume NN] bracketed
		bracketRe := regexp.MustCompile(`(?i)\s*\[(?:Volume|Vol\.?)\s*(\d+(?:\.\d+)?)\]`)
		if m := bracketRe.FindStringSubmatch(info.Title); m != nil && info.SeriesPosition == 0 {
			if v, err := strconv.ParseFloat(m[1], 64); err == nil {
				info.SeriesPosition = v
				info.Title = bracketRe.ReplaceAllString(info.Title, "")
				if info.Confidence < 55 {
					info.Confidence = 55
				}
			}
		}
	}

	// Post-process: clean up title after volume removal
	if info.Title != "" {
		// Remove trailing comma, dot, semicolon from volume removal artifacts
		info.Title = strings.TrimRight(info.Title, " ,;.")
		// Remove empty brackets
		info.Title = regexp.MustCompile(`\[\s*\]`).ReplaceAllString(info.Title, "")
		info.Title = strings.TrimSpace(info.Title)
		// Strip [PZG] or similar release group tags from start
		info.Title = regexp.MustCompile(`^\[[A-Z0-9]+\]\s*`).ReplaceAllString(info.Title, "")
		info.Title = strings.TrimSpace(info.Title)
		// Strip (Audiobook) (Unabridged) etc from end
		info.Title = regexp.MustCompile(`(?i)\s*\((?:Audiobook|Unabridged|Unabr)\)`).ReplaceAllString(info.Title, "")
		info.Title = regexp.MustCompile(`(?i)\s*\{[^}]*\}`).ReplaceAllString(info.Title, "") // {narrator} tags
		info.Title = strings.TrimSpace(info.Title)
	}

	// Post-process: extract title from "Series NN - Title" pattern
	if info.Series == "" && info.Title != "" {
		seriesRe := regexp.MustCompile(`^(\w+)\s+(\d+(?:\.\d+)?)\s+[-–—]\s+(.+)$`)
		if m := seriesRe.FindStringSubmatch(info.Title); m != nil {
			info.Series = m[1]
			if v, err := strconv.ParseFloat(m[2], 64); err == nil {
				info.SeriesPosition = v
			}
			info.Title = m[3]
			if info.Confidence < 65 {
				info.Confidence = 65
			}
		}
	}

	// Post-processing
	if info.Author != "" {
		info.Author = models.NormalizeAuthor(info.Author)
	}

	// Extract parent context
	parseParentContext(info, parentName)

	// Extract ASIN from anywhere
	extractASIN(clean, info)

	return info
}

// heuristicParse handles non-standard patterns that regex misses.
func heuristicParse(clean string, info *models.ParsedInfo) {
	// Pattern: "Series Name (Author)" — author in last parenthetical
	if lastOpen := strings.LastIndex(clean, "("); lastOpen >= 0 {
		lastClose := strings.LastIndex(clean, ")")
		if lastClose > lastOpen {
			before := strings.TrimSpace(clean[:lastOpen])
			parenContent := strings.TrimSpace(clean[lastOpen+1 : lastClose])

			// Strip all parentheticals from before
			before = stripAllParens(before)

			if !strings.Contains(before, " - ") && isAuthorish(parenContent) {
				info.Author = parenContent
				info.Series = before // Treat the "Series (Author)" format
				if info.Series != "" && !strings.ContainsAny(before, " -–—") {
					info.Confidence = maxInt(info.Confidence, 75)
				}
				return
			} else {
				// Author - Title (publisher info)
				clean = before
			}
		} else {
			clean = stripAllParens(clean)
		}
	} else {
		clean = stripAllParens(clean)
	}

	// Pattern: "Author - Title" or "Title - Author" via heuristics
	if strings.Contains(clean, " - ") {
		idx := strings.Index(clean, " - ")
		left := strings.TrimSpace(clean[:idx])
		right := strings.TrimSpace(clean[idx+3:])

		// Check for reverse pattern (Title - Author)
		if isAuthorName(right) && !isAuthorName(left) {
			info.Author = right
			info.Title = left
			info.Confidence = maxInt(info.Confidence, 70)
			return
		}

		// Standard: Author - Title
		info.Author = left
		info.Title = stripAllParens(right)
		info.Confidence = maxInt(info.Confidence, 65)
		return
	}

	// Fallback: use entire string as title
	if info.Title == "" {
		info.Title = clean
		if info.Confidence < 30 {
			info.Confidence = 30
		}
	}
}

// regexParse tries structured regex patterns against the clean name.
func regexParse(clean string, info *models.ParsedInfo) {
	for _, p := range regexPatterns {
		match := p.re.FindStringSubmatch(clean)
		if match == nil {
			continue
		}

		// Only override heuristic results if regex confidence is higher
		regexConf := 0
		if c, ok := p.mapping["_conf"]; ok {
			regexConf = c
		}
		if regexConf > info.Confidence {
			// Clear previous results for a clean match
			info.Title = ""
			info.Author = ""
			info.Series = ""
			info.SeriesPosition = 0
		}

		for fieldName, groupIdx := range p.mapping {
			if fieldName == "_conf" {
				info.Confidence = maxInt(info.Confidence, groupIdx)
			} else if groupIdx < len(match) {
				value := strings.TrimSpace(match[groupIdx])
				switch fieldName {
				case "title":
					if info.Title == "" {
						info.Title = value
					}
				case "author":
					if info.Author == "" {
						info.Author = value
					}
				case "series":
					if info.Series == "" {
						info.Series = value
					}
				case "series_position":
					if info.SeriesPosition == 0 {
						if v, err := strconv.ParseFloat(value, 64); err == nil {
							info.SeriesPosition = v
						}
					}
				case "asin":
					if info.ASIN == "" {
						info.ASIN = value
					}
				case "year":
					if info.Year == 0 {
						if v, err := strconv.Atoi(value); err == nil {
							info.Year = v
						}
					}
				}
			}
		}
		break
	}
}

// parseParentContext extracts series/author from the parent directory name.
func parseParentContext(info *models.ParsedInfo, parentName string) {
	if parentName == "" {
		return
	}

	// Try "Author - ..." pattern from parent for author
	if info.Author == "" {
		parentClean := stripAllParens(parentName)
		re := regexp.MustCompile(`^([^-–—]+?)\s*[-–—]\s+`)
		if m := re.FindStringSubmatch(parentClean); m != nil {
			info.Author = models.NormalizeAuthor(strings.TrimSpace(m[1]))
			if info.Confidence < 40 {
				info.Confidence = 40
			}
		}
	}

	// Try "Series (Author)" pattern from parent for series/author
	if info.Series == "" || info.Author == "" {
		if lastOpen := strings.LastIndex(parentName, "("); lastOpen >= 0 {
			lastClose := strings.LastIndex(parentName, ")")
			if lastClose > lastOpen {
				before := strings.TrimSpace(parentName[:lastOpen])
				parenContent := strings.TrimSpace(parentName[lastOpen+1 : lastClose])
				if isAuthorish(parenContent) && !strings.Contains(before, " - ") {
					if info.Author == "" {
						info.Author = models.NormalizeAuthor(parenContent)
						if info.Confidence < 50 {
							info.Confidence = 50
						}
					}
					if info.Series == "" && before != "" {
						info.Series = before
						if info.Confidence < 50 {
							info.Confidence = 50
						}
					}
					return
				}
			}
		}
	}

	// Look for series keywords in parent name
	keywordRe := regexp.MustCompile(`(?i)\s+(Series|Saga|Trilogy|Cycle|Chronicles|books\s+\d+)`)
	keywordMatch := keywordRe.FindStringIndex(parentName)

	if keywordMatch != nil && info.Series == "" {
		before := strings.TrimSpace(parentName[:keywordMatch[0]])
		words := strings.Fields(before)

		if len(words) > 0 {
			// Don't extract author from parent if before looks like title (starts with The/A/An)
			if info.Author == "" && len(words) >= 3 && len(words) <= 5 {
				lower := strings.ToLower(before)
				if !strings.HasPrefix(lower, "the ") && !strings.HasPrefix(lower, "a ") && !strings.HasPrefix(lower, "an ") {
					potentialAuthor := words[0] + " " + words[1]
					if isAuthorName(potentialAuthor) {
						info.Author = potentialAuthor
						if info.Confidence < 35 {
							info.Confidence = 35
						}
						info.Series = strings.Join(words[2:], " ")
						if info.Confidence < 40 {
							info.Confidence = 40
						}
						return
					}
				}
			}

			// Fallback: use entire before as series
			info.Series = before
			if info.Confidence < 40 {
				info.Confidence = 40
			}
		}
	}
}

// extractASIN finds an ASIN anywhere in the string.
func extractASIN(clean string, info *models.ParsedInfo) {
	if info.ASIN != "" {
		return
	}
	asinRe := regexp.MustCompile(`\[([A-Z0-9]{10})\]$`)
	if m := asinRe.FindStringSubmatch(clean); m != nil {
		info.ASIN = m[1]
		if info.Confidence < 60 {
			info.Confidence = 60
		}
	}
}

// isAuthorish checks if a string looks like a person's name (heuristic).
func isAuthorish(s string) bool {
	s = strings.TrimSpace(s)
	if s == "" {
		return false
	}
	words := strings.Fields(s)
	if len(words) < 1 || len(words) > 5 {
		return false
	}
	for _, w := range words {
		if len(w) > 15 {
			return false
		}
	}
	lower := strings.ToLower(s)
	for _, na := range nonAuthorWords {
		if strings.Contains(lower, na) {
			return false
		}
	}
	return true
}

// isAuthorName checks if the string IS an author name (stricter than isAuthorish).
// Used for reverse pattern detection (Title - Author).
func isAuthorName(s string) bool {
	s = strings.TrimSpace(s)
	if s == "" {
		return false
	}
	words := strings.Fields(s)
	if len(words) < 1 || len(words) > 4 {
		return false
	}
	lower := strings.ToLower(s)
	if strings.HasPrefix(lower, "the ") || strings.HasPrefix(lower, "a ") || strings.HasPrefix(lower, "an ") {
		return false
	}
	// Reject strings containing digits (not author names)
	for _, w := range words {
		for _, c := range w {
			if c >= '0' && c <= '9' {
				return false
			}
		}
	}
	titleWordCount := 0
	for _, w := range words {
		lw := strings.ToLower(w)
		if titleWords[lw] {
			titleWordCount++
		}
	}
	if len(words) > 2 && float64(titleWordCount)/float64(len(words)) > 0.3 {
		return false
	}
	for _, w := range words {
		if len(w) > 15 {
			return false
		}
	}
	return true
}

// isTitleLike checks if a string looks like a book title (not an author).
// Used to prevent treating book titles as author names.
func isTitleLike(s string) bool {
	s = strings.TrimSpace(s)
	if s == "" {
		return false
	}
	lower := strings.ToLower(s)
	// Titles often start with articles
	if strings.HasPrefix(lower, "the ") || strings.HasPrefix(lower, "a ") || strings.HasPrefix(lower, "an ") {
		return true
	}
	// Common title keywords
	titleKeywords := []string{
		"chronicles", "saga", "trilogy", "cycle", "series",
		"of ", "and ", "in ", "on ", "at ", "the ",
	}
	keywordCount := 0
	for _, kw := range titleKeywords {
		if strings.Contains(lower, kw) {
			keywordCount++
		}
	}
	words := strings.Fields(s)
	// Multi-word strings with many small words are likely titles
	if len(words) >= 4 {
		return true
	}
	// 3-word strings with at least one preposition/article likely title
	if len(words) == 3 && keywordCount >= 1 {
		return true
	}
	return keywordCount >= 2
}

// stripAllParens removes all parenthetical groups from a string.
func stripAllParens(s string) string {
	for {
		start := strings.Index(s, "(")
		if start < 0 {
			break
		}
		end := strings.Index(s[start:], ")")
		if end < 0 {
			break
		}
		end += start
		s = strings.TrimSpace(s[:start] + s[end+1:])
	}
	fields := strings.Fields(s)
	return strings.Join(fields, " ")
}

// StripExt removes the file extension from a path.
func StripExt(path string) string {
	ext := filepath.Ext(path)
	return strings.TrimSuffix(path, ext)
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Exported checks for use by the organizer.
// IsAuthorLike returns true if the string looks like an author name.
func IsAuthorLike(s string) bool { return isAuthorName(s) }

// IsTitleLike returns true if the string looks like a book title.
func IsTitleLike(s string) bool { return isTitleLike(s) }

// IsAuthorish returns true if the string might be an author name (loose check).
func IsAuthorish(s string) bool { return isAuthorish(s) }
